### PR TITLE
osd: Replacing friend declerations for passkey interface

### PR DESCRIPTION
**Pull request by Matan-B**
removing friend decleration from PrimaryLogScrub and ReplicaReservations

Signed-off-by: Matan Breizman <Matan.Brz@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/PG.h [hunk 1]
```diff
@@ -161,12 +161,21 @@ class PGRecoveryStats {
...
 
+class ScrubberPasskey {
+
+private:
+  friend class Scrub::ReplicaReservations;
+  friend class PrimaryLogScrub;
+  ScrubberPasskey() {}
+  ScrubberPasskey(const ScrubberPasskey&) = default;
+  ScrubberPasskey& operator=(const ScrubberPasskey&) = delete;
+
```
Comment (ronen-fr) [inline src/osd/PG.h:172]: rm the extra line in this case

File: src/osd/PG.h [hunk 2]
```diff
@@ -161,12 +161,21 @@ class PGRecoveryStats {
  *
  */
 
+class ScrubberPasskey {
+
```
Comment (ronen-fr) [inline src/osd/PG.h:165]: rm the extra line in this case

File: src/osd/PG.h [hunk 3]
```diff
@@ -161,12 +161,21 @@ class PGRecoveryStats {
  *
  */
 
+class ScrubberPasskey {
```
Comment (ronen-fr) [inline src/osd/PG.h:164]: Add a short comment to suggest use case.
Something like
"facilitating scrub-related object access to private PG data"

File: src/osd/PrimaryLogScrub.cc [hunk 4]
```diff
@@ -43,7 +43,7 @@ bool PrimaryLogScrub::get_store_errors(const scrub_ls_arg_t& arg,
 
 void PrimaryLogScrub::_scrub_finish()
 {
-  auto& info = m_pg->info;  ///< a temporary alias
+  auto& info{m_pg->get_pg_info(ScrubberPasskey())};  ///< a temporary alias
```
Comment (ronen-fr) [inline src/osd/PrimaryLogScrub.cc:46]: {}

Comment (ronen-fr) [reply src/osd/PrimaryLogScrub.cc:46]: to clarify that you are creating an object, not calling a function.

File: src/osd/pg_scrubber.h [hunk 5]
```diff
@@ -52,7 +52,10 @@ class ReplicaReservations {
 
   /// notify the scrubber that we have failed to reserve replicas' resources
   void send_reject();
-
+  
+  /// ScrubberPasskey Members
```
Comment (ronen-fr) [inline src/osd/pg_scrubber.h:56]: The comment is irrelevant here. We do not care about the mechanism at this specific point - just what this member does.

**Review by ronen-fr — CHANGES_REQUESTED** [review]
LGTM apart from some minor comments

**Review by ronen-fr — APPROVED** [review]

**Comment by ronen-fr** [comment]
Pls push it into 'ci', to make sure it compiles on all platforms.
Regarding Teuthology tests: not really important, as this a "compilation checked" change,
but do test with, e.g. rados::scrub

**Review by jdurgin — APPROVED** [review]
Nice cleanup! Always good to minimize the interface exposed, especially for an overly broad class like PG!

**Comment by ronen-fr** [comment]
Compilation-verified change. Still:
Passed Teu testing:
Runs:
original: yuriw-2021-09-09_21:20:54-rados-wip-yuri-testing-2021-09-09-1227-distro-basic-smithi (9F, 16D)
rerun: yuriw-2021-09-10_14:25:47-rados-wip-yuri-testing-2021-09-09-1227-distro-basic-smithi (4F, 1D)

Failures:

#6382926: “cephadm/test_dashboard_e2e.sh” - dashboard related failure. Seems to be https://tracker.ceph.com/issues/52417
#6382927: mgr/rook: might be https://tracker.ceph.com/issues/52321
#6382940: ceph_test_objectstore crash. Opened https://tracker.ceph.com/issues/52576
#6382942: seems the same as #6382926

The dead job: test issue (reimaging)

**Comment by ronen-fr** [comment]
Extra round of tests (thrash/scrub: all OK):
http://pulpito.front.sepia.ceph.com/matan-2021-09-12_09:32:47-rados:thrash-wip-matanb-scrub-passkey-distro-basic-smithi/

**Pull request merged**